### PR TITLE
Added Type prop to Input-Field 

### LINKED ISSUE #2237: Forms - Number Field
# Feature Request

## Description

Create a first class number input form field.

### PR DESCRIPTION
**Pull request by supreethmr**
### Summary
Resolves #2237 

### Additional Details
I have also added isLegendHidden prop to Form-Fieldset to fix issue : https://github.com/cerner/terra-clinical/issues/461

PR : https://github.com/cerner/terra-clinical/pull/481 requires isLegendHidden prop to be in Form-Fieldset so that it can be passed through the onset picker to maintain the consistency

Thanks for contributing to Terra. 
@cerner/terra

### CODE REVIEW

File: packages/terra-form-fieldset/src/Fieldset.jsx [hunk 1]
```diff
@@ -29,15 +29,20 @@ const propTypes = {
    * Determines whether the fieldset is required.
    */
   required: PropTypes.bool,
+  /**
+   * Whether or not the legend is visible. Use this props to hide a legend while still creating it on the DOM for accessibility.
```
Comment (nramamurth) [inline packages/terra-form-fieldset/src/Fieldset.jsx]: nit: prop*

File: packages/terra-form-fieldset/src/terra-dev-site/test/form-fieldset/FieldsetWithHiddenLegend.test.jsx [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+import Field from 'terra-form-field';
+import Input from 'terra-form-input';
+import Fieldset from '../../../Fieldset';
+
+const fieldsetwithhiddenlegend = () => (
+  <Fieldset
+    type="checkbox"
```
Comment (nramamurth) [inline packages/terra-form-fieldset/src/terra-dev-site/test/form-fieldset/FieldsetWithHiddenLegend.test.jsx]: I don't think that having `type`, `name`, `value` and `error` here are adding anything to the use-case.

File: packages/terra-form-fieldset/tests/wdio/fieldset-spec.js [hunk 3]
```diff
@@ -17,4 +17,11 @@ describe('Fieldset', () => {
       },
     });
   });
+
+  describe('Fieldset Populated with Hidden Legend', () => {
+    before(() => browser.url('/#/raw/tests/terra-form-fieldset/form-fieldset/fieldset-with-hidden-legend'));
+
+    Terra.should.beAccessible({ viewports });
```
Comment (nramamurth) [inline packages/terra-form-fieldset/tests/wdio/fieldset-spec.js]: Let's use `validateElement()` here.

File: packages/terra-form-input/src/terra-dev-site/doc/example/InputField.jsx [hunk 4]
```diff
@@ -6,10 +6,10 @@ const DefaultInputField = () => (
     inputId="profile-name"
     label="Profile Name"
     help="Note: This can not be changed in the future"
+    type="text"
```
Comment (nramamurth) [inline packages/terra-form-input/src/terra-dev-site/doc/example/InputField.jsx]: This is a non-passive change, methinks. For the existing users who have the current code something on the lines of snippet below, this change breaks the functionality because the default value of `undefined` overrides the existing value that is set. 

```
<InputField
  inputId="numeric-input"
  label="Numeric Value"
  help="Note: This can not be changed in the future"
  inputAttrs={{
    placeholder: 'Enter Digits',
    name: 'numeric',
    type: 'number'
  }}
/>
```

File: packages/terra-form-input/tests/jest/InputField.test.jsx [hunk 5]
```diff
@@ -20,8 +20,8 @@ it('should render a InputField with props', () => {
       hideRequired
       inputAttrs={{
         name: 'test',
-        type: 'number',
       }}
+      type="number"
```
Comment (nramamurth) [inline packages/terra-form-input/tests/jest/InputField.test.jsx]: Why was this changed? It is valid to have `type` as part of `inputAttrs` imo.

File: packages/terra-form-input/src/InputField.jsx [hunk 6]
```diff
@@ -163,6 +169,7 @@ const InputField = (props) => {
       <Input
         {...inputAttrs}
         id={inputId}
+        type={type}
```
Comment (nramamurth) [inline packages/terra-form-input/src/InputField.jsx]: This is a non-passive change, methinks. For the existing users who have the current code something on the lines of snippet below, this change breaks the functionality because the default value of `undefined` overrides the existing value that is set. 

```
<InputField
  inputId="numeric-input"
  label="Numeric Value"
  help="Note: This can not be changed in the future"
  inputAttrs={{
    placeholder: 'Enter Digits',
    name: 'numeric',
    type: 'number'
  }}
/>
```

File: packages/terra-form-input/tests/jest/InputField.test.jsx (hunk 5 above)
Comment (supreethmr) [reply packages/terra-form-input/tests/jest/InputField.test.jsx]: To keep it align with the code changes made for InputField. Now Input Field has type as first class prop. Since type prop will be having the priority over type specified within Inputattrs modified this example.

File: packages/terra-form-fieldset/src/Fieldset.jsx (hunk 1 above)
Comment (supreethmr) [reply packages/terra-form-fieldset/src/Fieldset.jsx]: done 👍

File: packages/terra-form-fieldset/src/terra-dev-site/test/form-fieldset/FieldsetWithHiddenLegend.test.jsx (hunk 2 above)
Comment (supreethmr) [reply packages/terra-form-fieldset/src/terra-dev-site/test/form-fieldset/FieldsetWithHiddenLegend.test.jsx]: yes we don't need them here It was copy paste mistake, [removed](https://github.com/cerner/terra-core/blob/e1f32ffcf116761ffddcc01deb5d3545ab8febca/packages/terra-form-fieldset/src/terra-dev-site/test/form-fieldset/FieldsetWithHiddenLegend.test.jsx#L7)

File: packages/terra-form-fieldset/tests/wdio/fieldset-spec.js (hunk 3 above)
Comment (supreethmr) [reply packages/terra-form-fieldset/tests/wdio/fieldset-spec.js]: [done](https://github.com/cerner/terra-core/blob/e1f32ffcf116761ffddcc01deb5d3545ab8febca/packages/terra-form-fieldset/tests/wdio/fieldset-spec.js#L24) 👍

File: packages/terra-form-input/src/InputField.jsx (hunk 6 above)
Comment (supreethmr) [reply packages/terra-form-input/src/InputField.jsx]: Added check for type prop : [`const inputType = (type) || inputAttrs.type;`](https://github.com/cerner/terra-core/blob/e1f32ffcf116761ffddcc01deb5d3545ab8febca/packages/terra-form-input/src/InputField.jsx#L152) . if type is not having any value assigned then the value of inputAttrs type will be assigned as type value for input.

File: packages/terra-form-input/src/terra-dev-site/doc/example/NumberInputField.jsx [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+import InputField from '../../../InputField';
+
+const NumberInputField = () => (
+  <InputField
+    inputId="numeric-input"
+    label="Numeric Value"
+    help="Note: This can not be changed in the future"
```
Comment (jmsv6d) [inline packages/terra-form-input/src/terra-dev-site/doc/example/NumberInputField.jsx]: I don't think that this help text makes sense for this example.

File: packages/terra-form-input/src/InputField.jsx [hunk 8]
```diff
@@ -143,6 +149,7 @@ const InputField = (props) => {
     }
   }
 
+  const inputType = (type) || inputAttrs.type;
```
Comment (jmsv6d) [inline packages/terra-form-input/src/InputField.jsx]: You shouldn't need the `()` around `type`.

**Comment by supreethmr** [comment]
> We may want to reconsider how we hide the legend. Using `display: none` removes the element from flow and will be ignored by screen readers. A better alternative is to use `terra-visually-hidden-text` or use the [visually hidden style directly](https://github.com/cerner/terra-core/blob/59bd29055cda77b0b12af8ee09db76d89497ef70/packages/terra-heading/src/Heading.module.scss#L35)

Awesome..! Thanks [visually hidden](https://github.com/cerner/terra-core/blob/8bdcebd9da82c7a46dd80a17f6c6b5a98fd38049/packages/terra-form-fieldset/src/Fieldset.module.scss#L32) works.

File: packages/terra-form-input/src/InputField.jsx (hunk 8 above)
Comment (supreethmr) [reply packages/terra-form-input/src/InputField.jsx]: [removed](https://github.com/cerner/terra-core/blob/8bdcebd9da82c7a46dd80a17f6c6b5a98fd38049/packages/terra-form-input/src/InputField.jsx#L152)

File: packages/terra-form-input/src/terra-dev-site/doc/example/NumberInputField.jsx (hunk 7 above)
Comment (supreethmr) [reply packages/terra-form-input/src/terra-dev-site/doc/example/NumberInputField.jsx]: [removed](https://github.com/cerner/terra-core/blob/8bdcebd9da82c7a46dd80a17f6c6b5a98fd38049/packages/terra-form-input/src/terra-dev-site/doc/example/NumberInputField.jsx)

File: packages/terra-form-input/tests/jest/InputField.test.jsx (hunk 5 above)
Comment (jmsv6d) [reply packages/terra-form-input/tests/jest/InputField.test.jsx]: We should have a test for both scenarios.

Comment (nramamurth) [reply packages/terra-form-input/tests/jest/InputField.test.jsx]: +1. Please add tests to cover all possible scenarios.

Comment (supreethmr) [reply packages/terra-form-input/tests/jest/InputField.test.jsx]: aded tests to cover inputAttrs and type prop scenario : [code](https://github.com/cerner/terra-core/pull/2337/commits/c150a05b7b97863658fa3ce7ea0a07b7158fe7a3)

File: packages/terra-form-fieldset/src/Fieldset.jsx [hunk 9]
```diff
@@ -52,7 +57,7 @@ const Fieldset = ({
 
   return (
     <fieldset {...customProps} className={fieldsetClasses}>
-      {legend && <legend {...legendAttrs} className={legendClasses}>{legend}</legend>}
+      {legend && <legend {...legendAttrs} className={cx([legendClasses, { 'legend-visually-hidden': isLegendHidden }])}>{legend}</legend>}
```
Comment (emilyrohrbough) [inline packages/terra-form-fieldset/src/Fieldset.jsx]: Lets move the legend-visually-hidden class to the legendClasses defined above on lines 53-56

Comment (supreethmr) [reply packages/terra-form-fieldset/src/Fieldset.jsx]: done 1a18a510b19e29aef23f0ae7303c2a7a390042d2

File: packages/terra-form-input/tests/jest/InputField.test.jsx [hunk 10]
```diff
@@ -41,6 +41,68 @@ it('should render a InputField with props', () => {
...
+      showOptional
+      value="Value"
+    />
+  );
+
+  const wrapper = shallow(textarea);
+  expect(wrapper).toMatchSnapshot();
+});
+
+it('should render a InputField with type specified through InpuAttributes', () => {
```
Comment (nramamurth) [inline packages/terra-form-input/tests/jest/InputField.test.jsx]: nit: *inputAttributes.

Comment (supreethmr) [reply packages/terra-form-input/tests/jest/InputField.test.jsx]: updated : 4eee8c3ec0e7fa14254d45cd5844087acffe4616
